mmcls/models/retriever/image2image.py

```diff
@@ -0,0 +1,315 @@
+            data. If None or no specified type, it will use
+            "ClsDataPreprocessor" as type. See :class:`ClsDataPreprocessor` for
+            more details. Defaults to None.
+        topk (int): Return the topk of the retrieval result. `-1` means
+            return all.
+            Defaults to -1.
+        prototype (Union[DataLoader, dict, str, torch.Tensor]): Database to be
+            retrieved. The following four types are supported.
+
+            - DataLoader: The original dataloader serves as the prototype.
+            - dict: The configuration to construct Dataloader.
+            - str: The path of the saved vector.
+            - torch.Tensor: The saved tensor whose dimension should be dim.
+
+        init_cfg (dict, optional): the config to control the initialization.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 image_encoder: Union[dict, List[dict]],
+                 head: Optional[dict] = None,
+                 pretrained: Optional[str] = None,
+                 similarity_fn: Union[str, Callable] = 'cosine_similarity',
+                 train_cfg: Optional[dict] = None,
+                 data_preprocessor: Optional[dict] = None,
+                 topk: int = -1,
+                 prototype: Union[DataLoader, dict, str, torch.Tensor] = None,
+                 init_cfg: Optional[dict] = None):
+
+        if pretrained is not None:
+            init_cfg = dict(type='Pretrained', checkpoint=pretrained)
+
+        if data_preprocessor is None:
+            data_preprocessor = {}
+        # The build process is in MMEngine, so we need to add scope here.
+        data_preprocessor.setdefault('type', 'mmcls.ClsDataPreprocessor')
+
+        if train_cfg is not None and 'augments' in train_cfg:
+            # Set batch augmentations by `train_cfg`
+            data_preprocessor['batch_augments'] = train_cfg
+
+        super(ImageToImageRetriever, self).__init__(
+            init_cfg=init_cfg, data_preprocessor=data_preprocessor)
+
+        self.image_encoder = MODELS.build(image_encoder)
+
+        if head is not None:
+            self.head = MODELS.build(head)
+        else:
+            self.head = None
+
+        self.similarity = similarity_fn
+
+        self.prototype = prototype
+        self.prototype_inited = False
+        self.prototype_vecs = None
+        self.topk = topk
+
+    @property
+    def similarity_fn(self):
+        """Returns a function that calculates the similarity."""
+        # If self.similarity_way is callable, return it directly
+        if isinstance(self.similarity, Callable):
+            return self.similarity
+
+        if self.similarity == 'cosine_similarity':
+            # a is a tensor with shape (N, C)
+            # b is a tensor with shape (M, C)
+            # "cosine_similarity" will get the matrix of similarity
+            # with shape (N, M)
+            return lambda a, b: torch.cosine_similarity(
```

Comment:
Sine matching returns `sorce` and `pred_label`, It needs to be clear whether the source is the bigger the better, or the smaller the better.